plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/cdk/server/core/internal/adapter/controllers/VagrantLaunchUtility.java

```diff
@@ -120,32 +126,104 @@ private ILaunchConfigurationWorkingCopy findLaunchConfig(IServer s, String launc
-			Map<String, String> env) throws IOException {
+	public static String[] call(String rootCommand, String[] args, File vagrantDir, Map<String, String> env)
+			throws IOException, TimeoutException {
+		return call(rootCommand, args, vagrantDir, env, 30000);
+	}
+
+	public static String[] call(String rootCommand, String[] args, File vagrantDir, Map<String, String> env,
+			int timeout) throws IOException, TimeoutException {
+
 		String[] envp = (env == null ? null : convertEnvironment(env));
 
-		List<String> result = new ArrayList<>();
 		List<String> cmd = new ArrayList<>();
 		cmd.add(rootCommand);
 		cmd.addAll(Arrays.asList(args));
-		Process p = Runtime.getRuntime().exec(cmd.toArray(new String[0]),
-				envp, vagrantDir);
-		BufferedReader buff = new BufferedReader(
-				new InputStreamReader(p.getInputStream()));
+		final Process p = Runtime.getRuntime().exec(cmd.toArray(new String[0]), envp, vagrantDir);
+
+		InputStream errStream = p.getErrorStream();
+		InputStream inStream = p.getInputStream();
+
+		StreamGobbler inGobbler = new StreamGobbler(inStream);
+		StreamGobbler errGobbler = new StreamGobbler(errStream);
+		
+		new Thread(inGobbler).start();
+		new Thread(errGobbler).start();
+
+		
+		Integer exitCode = ThreadUtils.runWithTimeout(timeout, new Callable<Integer>() {
+			@Override
+		 	public Integer call() throws Exception {
+				return p.waitFor();
+			}
+		});
+		
 		try {
-			if (p.waitFor() == 0) {
-				String line;
-				while ((line = buff.readLine()) != null) {
-					result.add(line);
-				}
-			} else {
-				return new String[0];
+			if( exitCode == null ) {
+				// Timeout reached
+				inGobbler.cancel();
+				errGobbler.cancel();
+				throw new TimeoutException(getTimeoutError(inGobbler.getLines(), errGobbler.getLines()));
 			}
-		} catch(InterruptedException ie) {
-			// ignore
+			
+			return inGobbler.getLines();
+		} finally {
+			inGobbler.cleanup();
+			errGobbler.cleanup();
 		}
-		return result.toArray(new String[0]);
 	}
 	
+	private static String getTimeoutError(String[] output, String[] err) {
+		StringBuffer msg = new StringBuffer();
+		msg.append("Process output:\n");
+		String[] lines = output;
+		for( int i = 0; i < lines.length; i++ ) {
+			msg.append("    ");
+			msg.append(lines[i]);
+		}
+		msg.append("\n\nProcess Error:\n");
+		lines = err;
+		for( int i = 0; i < lines.length; i++ ) {
+			msg.append("    ");
+			msg.append(lines[i]);
+		}
+		return msg.toString();
+	}
 	
+	private static class StreamGobbler implements Runnable {
+		private Scanner inScanner;
+		private ArrayList<String> lines;
+		private boolean canceled = false;
+		private BufferedInputStream is;
+		public StreamGobbler(InputStream inStream) {
+			lines = new ArrayList<String>();
+			is = new BufferedInputStream(inStream);
+			inScanner = new Scanner(is);
+		}
+
+		@Override
+		public void run() {
+			while (inScanner.hasNextLine() && !isCanceled()) {
+				String line = inScanner.nextLine();
+				lines.add(line);
```

Comment:
When cancel() is called, this thread can be right before adding line to lines, and that may cause concurrent exception while the other thread calls getLines() or works with list if getLines() will return list. I suggest wrap it with

		synchronized (this) {
			if(!isCanceled()) {
				lines.add(line);
			}
		}

and call cancel() as the first command in getLines().